lib/PreviewProvider.js

```diff
@@ -0,0 +1,226 @@
+        if (!test(item)) {
+          result = false;
+        }
+      });
+      return result;
+    };
+  },
+
+  /**
+    * Santize the URL to remove any unwanted or sensitive information about the link
+    */
+  _sanitizeURL(url) {
+    if (!url) {
+      return "";
+    }
+
+    let queryParams = {};
+    let safeQueryParam = "";
+    let newURL = new URL(url);
+
+    // extract and parse the query parameters, if any
+    if (newURL.search.length !== 0) {
+      let tmp = new Map(newURL.search.slice(1).split("&").map(pair => pair.split("=").map(decodeURIComponent)));
+      tmp.forEach((val, key) => queryParams[key] = val);
+
+      // filter out the allowed query params and update the query string
+      Object.keys(queryParams)
+        .filter(param => ALLOWED_QUERY_PARAMS.has(param))
+        .forEach(param => {
+          safeQueryParam = safeQueryParam + param + "=" + queryParams[param] + "&";
+        });
+      newURL.search = safeQueryParam.slice(0, safeQueryParam.length - 1);
+    }
+
+    // remove extra slashes, and instances of '/../' and '/./' then construct back a safe pathname
+    if (newURL.pathname) {
+      let safePathItems = newURL.pathname.split("/").filter(item => item.replace(/\.{1,}/, ""));
+      let safePath = "/";
+      newURL.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url contains sensitive information, remove it
+    if (newURL.username) {
+      newURL.username = "";
+    }
+    if (newURL.password) {
+      newURL.password = "";
+    }
+
+    newURL.hash = "";
+    return newURL.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createCacheKey(url) {
+    url = new URL(url);
+    let key = url.host.replace(/www\.?/, "");
+    key = key + url.pathname;
+    return key.toString();
+  },
+
+  /**
+    * Canonicalize urls by sanitizing them, then deduping them
+    */
+  _uniqueLinks(links) {
+    let dedupedSites = new Map();
+    links.forEach(site => {
+      site.sanitizedURL = this._sanitizeURL(site.url);
+      site.cacheKey = this._createCacheKey(site.url);
+      if (!dedupedSites.has(site.cacheKey)) {
+        dedupedSites.set(site.cacheKey, site);
+      }
+    });
+    return Array.from(dedupedSites.values());
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    let sites = this._uniqueLinks(links.filter(this._URLFilter(URL_FILTERS)));
+    sites.forEach(link => {
+      if (link && ss.storage.embedlyData[link.cacheKey] && link.cacheKey === ss.storage.embedlyData[link.cacheKey].cacheKey) {
+        cachedLinks.push(ss.storage.embedlyData[link.cacheKey]);
+      }
+    });
+
+    return cachedLinks;
+  },
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let sites = this._uniqueLinks(links.filter(this._URLFilter(URL_FILTERS)));
```

Comment:
We're doing this twice, here and on line 144.  Is it possible to do this once and then pass that dataset around?